internal/graph/rate_limited_check_resolver.go

```diff
@@ -0,0 +1,176 @@
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+// RateLimitedCheckResolverConfig encapsulates configuration for rate limited check resolver
+type RateLimitedCheckResolverConfig struct {
+	TimerTickerFrequency time.Duration
+	MediumPriorityLevel  uint32
+	MediumPriorityShaper uint32
+	LowPriorityLevel     uint32
+	LowPriorityShaper    uint32
+}
+
+// RateLimitedCheckResolver will prioritize requests with fewer dispatches over
+// requests with more dispatches.
+// Initially, request's dispatches will not be throttled and will be processed
+// immediately. As the number of dispatches increases (to beyond the config's MediumPriorityLevel),
+// selectively number (configure via config's MediumPriorityShaper) will be throttled by placing the
+// dispatch in the medium priority queue. One item from the medium priority queue will be processed
+// every ticker across the entire server. Request's dispatches will gradually be more frequently throttled as
+// the number of dispatches increase until every dispatch will be throttled by being placed in the medium priority
+// queue. When the number of request dispatches is above the LowPriorityLevel, the dispatches are placed
+// in the low priority queue. One item form the low priority queue will be processed every Nth ticker
+// (configured via config's LowPriorityShaper parameter).
+// This allows a check / list objects request to be gradually throttled.
+type RateLimitedCheckResolver struct {
+	delegate         CheckResolver
+	config           RateLimitedCheckResolverConfig
+	ticker           *time.Ticker
+	lowPriorityQueue chan bool
+	medPriorityQueue chan bool
+	done             chan bool
+
+	// these are helper config value to reduce calculation needed for medium shaper
+	mediumPriorityLevel2      uint32
+	mediumPriorityLevel3      uint32
+	mediumPriorityLevel4      uint32
+	mediumPriorityShaperFreq2 uint32
+	mediumPriorityShaperFreq3 uint32
+}
+
+var _ CheckResolver = (*RateLimitedCheckResolver)(nil)
+
+var (
+	rateLimitedCheckResolverDelayMsHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "rate_limited_check_resolver_delay_ms",
+		Help:                            "Time spent waiting for rate limited check resolver",
+		Buckets:                         []float64{1, 3, 5, 10, 25, 50, 100, 1000, 5000}, // Milliseconds. Upper bound is config.UpstreamTimeout.
+		NativeHistogramBucketFactor:     1.1,
+		NativeHistogramMaxBucketNumber:  100,
+		NativeHistogramMinResetDuration: time.Hour,
+	}, []string{"grpc_service", "grpc_method", "queue_name"})
+)
+
+func NewRateLimitedCheckResolver(
+	config RateLimitedCheckResolverConfig) *RateLimitedCheckResolver {
+	ticker := time.NewTicker(config.TimerTickerFrequency)
```

Comment:
nit: you don't need this ticker to be a struct field, it can just be a local variable inside `runTicker` since that is the only place where it is used

Reply:
updated